controller/volume_controller.go

```diff
@@ -2367,52 +2372,49 @@ func (vc *VolumeController) processMigration(v *longhorn.Volume, es map[string]*
 		return nil
 	}
 
-	if len(es) > 2 {
-		return fmt.Errorf("BUG: volume %v: more than two engines exists", v.Name)
-	}
-
+	// the only time there should be more then 1 engines is when we are migrating or upgrading
+	// if there are more then 1 and we no longer have a migration id set we can cleanup the extra engine
 	if !vc.isVolumeMigrating(v) {
-		if len(es) != 2 {
+		if len(es) < 2 {
 			return nil
 		}
 
-		// rollback migration
-		if _, err := vc.getCurrentEngineAndCleanupOthers(v, es); err != nil {
-			return err
+		// in the case of a confirmation we need to switch the v.Status.CurrentNodeID to v.Spec.NodeID
+		// so that currentEngine becomes the migration engine
+		if v.Status.CurrentNodeID != v.Spec.NodeID {
+			vc.logger.Infof("volume migration complete switching current node id from %v to %v", v.Status.CurrentNodeID, v.Spec.NodeID)
+			v.Status.CurrentNodeID = v.Spec.NodeID
 		}
-		return nil
-	}
 
-	// confirm migration
-	if v.Spec.MigrationNodeID == v.Spec.NodeID {
-		// must be set after migration preparation done
+		// current engine is based on the v.Status.CurrentNodeID
+		// so in the case of a rollback all we have to do is cleanup the migration engine
 		currentEngine, err := vc.getCurrentEngineAndCleanupOthers(v, es)
 		if err != nil {
 			return err
 		}
 
+		// cleanupCorruptedOrStaleReplicas() will take care of old replicas
 		if err := vc.switchActiveReplicas(rs, func(r *longhorn.Replica, engineName string) bool {
 			return r.Spec.EngineName == engineName
 		}, currentEngine.Name); err != nil {
 			return err
 		}
 
-		v.Status.OwnerID = v.Spec.NodeID
-		v.Spec.MigrationNodeID = ""
-		if _, err := vc.ds.UpdateVolume(v); err != nil {
-			return err
-		}
-
-		// cleanupCorruptedOrStaleReplicas() will take care of old replicas
-		vc.logger.Infof("Migration to node %v has been confirmed", v.Spec.NodeID)
 		return nil
 	}
 
-	currentEngine, migrationEngine, err := vc.getCurrentEngineAndExtra(v, es)
+	// volume is migrating
+	currentEngine, extras, err := vc.getCurrentEngineAndExtras(v, es)
 	if err != nil {
 		return err
 	}
 
+	var migrationEngine *longhorn.Engine
+	for _, extra := range extras {
```

Comment:
If there are more than 2 engines, can we continue migration? Allowing cleanup is good in this case but I think we should prevent the actual migration.

Reply:
One running volume having more than 2 engines is unexpected. I prefer to error out...

Reply:
If we just error out, how will the volume ever recover from that situation? -> endless error loop
Just erroring out in this case is not a desired behavior, since the goal should always be to be resilient and recover from failure states.

Reply:
I added some cleanup code, to take care of volumes in an invalid state for migration.